File: excelparser.go
package main

import (
	"database/sql"
	"fmt"
	"log"
	"path/filepath"
	"strings"
	"time"

	_ "github.com/mattn/go-sqlite3"
	"github.com/xuri/excelize/v2"
)

type Anketa struct {
	fullname   string
	previous   string
	birthday   string
	birthplace string
	country    string
	snils      string
	inn        string
	education  string
}

type Check struct {
	workplace  string
	cronos     string
	cros       string
	document   string
	debt       string
	bankruptcy string
	bki        string
	affilation string
	internet   string
	pfo        bool
	addition   string
	conclusion int
	officer    string
	deadline   time.Time
}

type Robot struct {
	inn        string
	employee   string
	terrorist  string
	mvd        string
	courts     string
	bankruptcy string
	bki        string
}

func excelParse(db *sql.DB, excelPath string, excelFile string) {
	queries := map[string]string{
		"updatePerson":      "UPDATE persons SET fullname = ?, previous = ?, birthday = ?, birthplace = ?, country = ?, snils = ?, inn = ?, education = ?, updated = ?, category_id = ?, region_id = ?, status_id = ? WHERE id = ?",
		"updateShortPerson": "UPDATE persons SET fullname = ?, birthday = ?, updated = ?, category_id = ?, region_id = ?, status_id = ? WHERE id = ?",
		"insertShortPerson": "INSERT INTO persons (fullname, birthday, created, category_id, region_id, status_id) VALUES (?, ?, ?, ?, ?, ?)",
		"insertPerson":      "INSERT INTO persons (fullname, previous, birthday, birthplace, country, snils, inn, education, created, category_id, region_id, status_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
		"insertCheck":       "INSERT INTO checks (workplace, cronos, cros, document, debt, bankruptcy, bki, affilation, internet, pfo, addition, conclusion, officer, deadline, person_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
		"insertRobot":       "INSERT INTO robots (inn, employee, terrorist, mvd, courts, bankruptcy, bki, deadline, person_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
	}

	stmts := make(map[string]*sql.Stmt)
	for key, query := range queries {
		stmt, err := db.Prepare(query)
		if err != nil {
			log.Fatal(err)
		}
		defer stmt.Close()
		stmts[key] = stmt
	}

	var candId int
	var fio string
	var anketa Anketa

	f, err := excelize.OpenFile(filepath.Join(excelPath, excelFile))
	if err != nil {
		log.Println(err)
		return
	}
	defer f.Close()

	if strings.HasPrefix(excelFile, "Заключение") {
		name, _ := f.GetCellValue("Лист1", "C6")
		if name == "" {
			return
		}
		anketa.fullname = strings.ToTitle(strings.Join(strings.Fields(name), " "))
		anketa.birthday, _ = parseDateCell(f, "Лист1", "C8")
		if err != nil {
			anketa.birthday = "2006-01-02"
		}
		anketa.previous, _ = f.GetCellValue("Лист1", "C7")

		if f.SheetCount > 1 {
			fio, err = f.GetCellValue("Лист2", "K1")
			if err != nil {
				log.Println(err)
				return
			}

			if fio == "ФИО" {
				name, _ := f.GetCellValue("Лист2", "K3")
				if name == "" {
					return
				}
				anketa.fullname = strings.ToTitle(strings.Join(strings.Fields(name), " "))
				anketa.previous, _ = f.GetCellValue("Лист2", "S3")
				anketa.birthday, _ = parseDateCell(f, "Лист2", "L3")
				if err != nil {
					anketa.birthday = "2006-01-02"
				}
				anketa.birthplace, _ = f.GetCellValue("Лист2", "M3")
				anketa.country, _ = f.GetCellValue("Лист2", "T3")
				anketa.snils, _ = f.GetCellValue("Лист2", "U3")
				anketa.inn, _ = f.GetCellValue("Лист2", "V3")
				anketa.education, _ = f.GetCellValue("Лист2", "W3")
			}
		}
		var check Check
		wp1, _ := f.GetCellValue("Лист1", "D11")
		wp2, _ := f.GetCellValue("Лист1", "D12")
		wp3, _ := f.GetCellValue("Лист1", "D13")
		check.workplace = fmt.Sprintf("%s; %s; %s", wp1, wp2, wp3)
		check.cronos, _ = f.GetCellValue("Лист1", "C14")
		check.cros, _ = f.GetCellValue("Лист1", "C16")
		check.document, _ = f.GetCellValue("Лист1", "C17")
		check.debt, _ = f.GetCellValue("Лист1", "C18")
		check.bankruptcy, _ = f.GetCellValue("Лист1", "C19")
		check.bki, _ = f.GetCellValue("Лист1", "C20")
		check.affilation, _ = f.GetCellValue("Лист1", "C21")
		check.internet, _ = f.GetCellValue("Лист1", "C22")
		check.officer, _ = f.GetCellValue("Лист1", "C25")
		check.addition, _ = f.GetCellValue("Лист1", "C28")

		decision, err := f.GetCellValue("Лист1", "C23")
		if err != nil {
			decision = "Согласовано"
		}
		row := db.QueryRow(
			"SELECT id FROM conclusions WHERE LOWER(conclusion) = $1",
			strings.ToLower(decision),
		)
		err = row.Scan(&check.conclusion)
		if err != nil {
			check.conclusion = 1
		}
		check.deadline = time.Now()
		poligraf, err := f.GetCellValue("Лист1", "C26")
		if err != nil {
			poligraf = "назначено"
		}
		if strings.ToLower(poligraf) == "назначено" || poligraf == "на испытательном сроке" {
			check.pfo = true
		} else {
			check.pfo = false
		}

		result := db.QueryRow(
			"SELECT id FROM persons WHERE fullname = ? AND birthday = ?",
			anketa.fullname, anketa.birthday,
		)
		err = result.Scan(&candId)
		if err != nil {
			if err == sql.ErrNoRows {
				result, err := stmts["insertPerson"].Exec(
					anketa.fullname, anketa.previous, anketa.birthday,
					anketa.birthplace, anketa.country, anketa.snils,
					anketa.inn, anketa.education, time.Now(),
					categoryId, regionId, statusId,
				)
				if err != nil {
					log.Println(err)
					return
				}
				id, err := result.LastInsertId()
				if err != nil {
					log.Println(err)
					return
				}
				candId = int(id)
			} else {
				log.Println(err)
				return
			}

		} else {
			if fio == "" {
				_, err := stmts["updatePerson"].Exec(
					anketa.fullname, anketa.previous, anketa.birthday, anketa.birthplace,
					anketa.country, anketa.snils, anketa.inn, anketa.education,
					time.Now(), categoryId, regionId, statusId, candId,
				)
				if err != nil {
					log.Println(err)
					return
				}
			} else {
				_, err := stmts["updateShortPerson"].Exec(
					anketa.fullname, anketa.birthday, time.Now(), categoryId,
					regionId, statusId, candId,
				)
				if err != nil {
					log.Println(err)
					return
				}
			}

		}

		_, err = stmts["insertCheck"].Exec(
			check.workplace, check.cronos, check.cros, check.document, check.debt,
			check.bankruptcy, check.bki, check.affilation, check.internet,
			check.pfo, check.addition, check.conclusion, check.officer, time.Now(), candId,
		)
		if err != nil {
			log.Println(err)
			return
		}

	} else {
		var robot Robot
		name, _ := f.GetCellValue("Лист1", "B4")
		anketa.fullname = strings.ToTitle(strings.Join(strings.Fields(name), " "))
		anketa.birthday, err = parseDateCell(f, "Лист1", "B5")
		if err != nil {
			anketa.birthday = "2006-01-02"
		}

		robot.employee, _ = f.GetCellValue("Лист1", "B27")
		robot.terrorist, _ = f.GetCellValue("Лист1", "B17")
		robot.inn, _ = f.GetCellValue("Лист1", "B18")
		bnkrpt, _ := f.GetCellValue("Лист1", "B20")
		bnkrpt2, _ := f.GetCellValue("Лист1", "B21")
		bnkrpt3, _ := f.GetCellValue("Лист1", "B22")
		robot.bankruptcy = fmt.Sprintf("%s, %s, %s", bnkrpt, bnkrpt2, bnkrpt3)
		robot.mvd, _ = f.GetCellValue("Лист1", "B23")
		robot.courts, _ = f.GetCellValue("Лист1", "B24")
		robot.bki, _ = f.GetCellValue("Лист1", "B25")

		result := db.QueryRow(
			"SELECT id FROM persons WHERE fullname = ? AND birthday = ?",
			anketa.fullname, anketa.birthday,
		)
		err = result.Scan(&candId)
		if err != nil {
			if err == sql.ErrNoRows {
				result, err := stmts["insertShortPerson"].Exec(
					anketa.fullname, anketa.birthday, time.Now(),
					categoryId, regionId, statusId,
				)
				if err != nil {
					log.Println(err)
					return
				}
				id, err := result.LastInsertId()
				if err != nil {
					log.Println(err)
					return
				}
				candId = int(id)
			} else {
				log.Println(err)
				return
			}

		} else {
			_, err := stmts["updateShortPerson"].Exec(
				anketa.fullname, anketa.birthday, time.Now(),
				categoryId, regionId, statusId, candId,
			)
			if err != nil {
				log.Println(err)
				return
			}
		}

		_, err = stmts["insertRobot"].Exec(robot.inn, robot.employee, robot.terrorist,
			robot.mvd, robot.courts, robot.bankruptcy, robot.bki, time.Now(), candId)
		if err != nil {
			log.Println(err)
			return
		}
	}
}

File: excelparser.py
import os
from datetime import date, datetime

import aiosqlite
from openpyxl import load_workbook

from config import Config


async def excel_to_db(excel_path, excel_file):
    excel = await screen_excel(excel_path, excel_file)
    if excel.get("resume"):
        excel["resume"] | {"category_id": 1, "status_id": 9, "region_id": 1}

        async with aiosqlite.connect(Config.DATABASE_URI) as db:
            async with db.execute(
                "SELECT * FROM persons WHERE fullname = ? AND birthday = ?",
                (excel["resume"]["fullname"], excel["resume"]["birthday"]),
            ) as cursor:
                result = await cursor.fetchone()
                person_id = result[0] if result else None

            if not person_id:
                await db.execute(
                    f"INSERT INTO persons ({','.join(excel['resume'].keys())}), created "
                    f"VALUES ({','.join(['?'] * len(excel['resume'].values()))}), ?",
                    tuple(excel["resume"].values()) + (datetime.now(),),
                )
                person_id = cursor.lastrowid
            else:
                await db.execute(
                    f"UPDATE persons SET {'=?,'.join(excel['resume'].keys())}=?, updated=? "
                    f"WHERE id = {person_id}",
                    tuple(excel["resume"].values()) + (datetime.now(),),
                )

            if excel.get("check"):
                await db.execute(
                    f"INSERT INTO checks ({','.join(excel['check'].keys())}, person_id) "
                    f"VALUES ({','.join(['?'] * len(excel['check'].values()))}, ?)",
                    tuple(excel["check"].values()) + (person_id,),
                )
            elif excel.get("robot"):
                await db.execute(
                    f"INSERT INTO robots ({','.join(excel['robot'].keys())}, person_id) "
                    f"VALUES ({','.join(['?'] * len(excel['robot'].values()))}, ?)",
                    tuple(excel["robot"].values()) + (person_id,),
                )

            await db.commit()


async def screen_excel(excel_path, excel_file):
    workbook = load_workbook(os.path.join(excel_path, excel_file), keep_vba=True)
    worksheet = workbook.worksheets[0]
    person = {}

    if excel_file.startswith("Заключение"):
        if len(workbook.sheetnames) > 1:
            sheet = workbook.worksheets[1]
            if str(sheet["K1"].value) == "ФИО":
                person.update({"resume": await get_resume(sheet)})
        person.update({"resume": await get_conclusion_resume(worksheet)})
        person.update({"check": await get_check(worksheet)})
    else:
        person.update({"resume": await get_robot_resume(worksheet)})
        person.update({"robot": await get_robot(worksheet)})

    workbook.close()
    return person


async def get_resume(sheet):
    resume = {
        "fullname": fullname_parser(str(sheet["K3"].value)),
        "birthday": datetime.strptime(sheet["L3"].value, "%d.%m.%Y").date()
        if sheet["L3"].value
        else date.today(),
        "birthplace": str(sheet["M3"].value).strip(),
        "country": str(sheet["T3"].value).strip(),
        "snils": str(sheet["U3"].value).strip(),
        "inn": str(sheet["V3"].value).strip(),
    }
    return resume


async def get_conclusion_resume(sheet):
    resumes = {
        "fullname": fullname_parser(sheet["C6"].value),
        "birthday": (sheet["C8"].value).date()
        if isinstance(sheet["L3"].value, datetime)
        else date.today(),
    }
    return resumes


async def get_robot_resume(sheet):
    resumes = {
        "fullname": fullname_parser(sheet["B4"].value),
        "birthday": datetime.strptime(sheet["B5"].value, "%d.%m.%Y").date()
        if sheet["B5"].value
        else date.today(),
    }
    return resumes


async def get_check(sheet):
    checks = {
        "workplace": f"{sheet['C11'].value} - {sheet['D11'].value}; {sheet['C12'].value} - "
        f"{sheet['D12'].value}; {sheet['C13'].value} - {sheet['D13'].value}",
        "cronos": f"{sheet['B14'].value}: {sheet['C14'].value}; {sheet['B15'].value}: "
        f"{sheet['C15'].value}",
        "cros": sheet["C16"].value,
        "document": sheet["C17"].value,
        "debt": sheet["C18"].value,
        "bankruptcy": sheet["C19"].value,
        "bki": sheet["C20"].value,
        "affilation": sheet["C21"].value,
        "internet": sheet["C22"].value,
        "pfo": True if sheet["C26"].value else False,
        "addition": sheet["C28"].value,
        "conclusion": await get_conclusion_id(sheet["C23"].value),
        "deadline": datetime.now(),
        "officer": sheet["C25"].value,
    }
    return checks


async def get_robot(sheet):
    robot = {
        "employee": sheet["B27"].value,
        "terrorist": sheet["B17"].value,
        "inn": sheet["B18"].value,
        "bankruptcy": f"{sheet['B20'].value}, {sheet['B21'].value}, {sheet['B22'].value}",
        "mvd": sheet["B23"].value,
        "courts": sheet["B24"].value,
        "bki": sheet["B25"].value,
        "deadline": datetime.now(),
    }
    return robot


async def get_conclusion_id(name):
    async with aiosqlite.connect(Config.DATABASE_URI) as conn:
        async with conn.execute(
            "SELECT * FROM conclusions WHERE LOWER (conclusion) = ?", (name.lower(),)
        ) as cursor:
            result = await cursor.fetchone()
            return result[0] if result else 1


def fullname_parser(fullname: str) -> str:
    return " ".join(filter(None, map(str.strip, fullname.split()))).upper()
